lte/gateway/c/session_manager/LocalSessionManagerHandler.cpp

```diff
@@ -139,58 +141,57 @@ bool LocalSessionManagerHandlerImpl::is_pipelined_restarted() {
   }
 
   if (current_epoch_ != epoch) {
-    MLOG(MDEBUG) << "Received stale Pipelined setup callback for " << epoch
-                 << ", current epoch is " << current_epoch_;
+    // This means that PipelineD has restarted since the initial Setup call was
+    // called
+    MLOG(MDEBUG) << "Received stale PipelineD setup callback for epoch: "
+                 << epoch << ", current epoch: " << current_epoch_;
+    return;
+  }
+  if (status.ok() && resp.result() == resp.OUTDATED_EPOCH) {
+    MLOG(MWARNING) << "PipelineD setup call has outdated epoch, abandoning.";
     return;
   }
 
+  // Cases for which we re-try the Setup call
   if (!status.ok()) {
-    MLOG(MERROR) << "Could not setup pipelined, rpc failed with: "
-                 << status.error_message() << ", retrying pipelined setup "
-                 << "for epoch " << epoch;
-
-  } else if (resp.result() == resp.OUTDATED_EPOCH) {
-    MLOG(MWARNING) << "Pipelined setup call has outdated epoch, abandoning.";
-    return;
+    MLOG(MERROR) << "Could not setup PipelineD, rpc failed with: "
+                 << status.error_message() << ", retrying PipelineD setup "
+                 << "for epoch: " << epoch;
   } else if (resp.result() == resp.FAILURE) {
-    MLOG(MWARNING) << "Pipelined setup failed, retrying pipelined setup "
-                      "after delay, for epoch "
-                   << epoch;
+    MLOG(MWARNING) << "PipelineD setup failed, retrying PipelineD setup "
+                   << "after delay, for epoch: " << epoch;
   }
 
   enforcer_->get_event_base().runInEventBaseThread([=] {
     enforcer_->get_event_base().timer().scheduleTimeoutFn(
-        std::move([=] {
-          auto session_map = session_store_.read_all_sessions();
-          enforcer_->setup(
-              session_map, epoch,
-              std::bind(
-                  &LocalSessionManagerHandlerImpl::handle_setup_callback, this,
-                  epoch, _1, _2));
-        }),
-        retry_timeout_);
+        std::move([=] { call_setup_pipelined(epoch); }), retry_timeout_ms_);
   });
 }
 
-bool LocalSessionManagerHandlerImpl::restart_pipelined(
+void LocalSessionManagerHandlerImpl::call_setup_pipelined(
     const std::uint64_t& epoch) {
   using namespace std::placeholders;
-  enforcer_->get_event_base().runInEventBaseThread([this, epoch]() {
-    auto session_map = session_store_.read_all_sessions();
-    enforcer_->setup(
-        session_map, epoch,
-        std::bind(
-            &LocalSessionManagerHandlerImpl::handle_setup_callback, this, epoch,
-            _1, _2));
-  });
-  return true;
+  if (is_setting_up_pipelined_) {
+    // Return if there is already a Setup call in progress
+    return;
+  }
+  is_setting_up_pipelined_ = true;
+
+  auto session_map = session_store_.read_all_sessions();
+  enforcer_->setup(
+      session_map, epoch,
+      std::bind(
+          &LocalSessionManagerHandlerImpl::handle_setup_callback, this, epoch,
+          _1, _2));
+  return;
```

Comment:
I understand that `is_setting_up_pipelined_==true` from line 185, until enforcer.setup calls its callback  `handle_setup_callback`.  But i see all that code belongs to the same event. 

How is it possible we may get another cal to `call_setup_pipelined`

Reply:
Multiple `call_setup_pipelined` get scheduled on the event loop since each of these three events does this:
1. Failure from `SetupPolicyFlows`
2. Failure from `SetupDefaultControllers`
3. A new epoch from ReportRuleStats
So we want to allow the scheduling, but only act on a new setup call if it's not being done already.